Make CUDA serde support for TP agent pluggable

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59788 Prepare for TensorPipe separating its CUDA-specific headers
* #59377 Move CUDA-related stuff of TP agent to separate file
* **#59376 Make CUDA serde support for TP agent pluggable**

This is an experiment. The end goal is to separate the CUDA-specific aspects of the TensorPipe agent so that they can be plugged "on top" of the CPU-only parts. This will then allow to move the TP agent to libtorch (because libtorch is split into a CPU and a CUDA part; now it's in libtorch_python), although unfortunately other conditions need to also be met for this to happen.

The only instance where we had CPU and CUDA logic within the same code, guarded by `#ifdef USE_CUDA`, is the serialization/deserialization code. I'm thus introducing a sort-of registry in order to "decentralize it". It's not a c10::Registry, because that's overkill (it uses an unordered_map, with strings as keys): here we can just use an array with integers as "keys".

Differential Revision: [D28796428](https://our.internmc.facebook.com/intern/diff/D28796428/)

lw committed:
Make CUDA serde support for TP agent pluggable

This is an experiment. The end goal is to separate the CUDA-specific aspects of the TensorPipe agent so that they can be plugged "on top" of the CPU-only parts. This will then allow to move the TP agent to libtorch (because libtorch is split into a CPU and a CUDA part; now it's in libtorch_python), although unfortunately other conditions need to also be met for this to happen.

The only instance where we had CPU and CUDA logic within the same code, guarded by `#ifdef USE_CUDA`, is the serialization/deserialization code. I'm thus introducing a sort-of registry in order to "decentralize it". It's not a c10::Registry, because that's overkill (it uses an unordered_map, with strings as keys): here we can just use an array with integers as "keys".

Differential Revision: [D28796428](https://our.internmc.facebook.com/intern/diff/D28796428/)

[ghstack-poisoned]

lw committed:
Update on "Make CUDA serde support for TP agent pluggable"


This is an experiment. The end goal is to separate the CUDA-specific aspects of the TensorPipe agent so that they can be plugged "on top" of the CPU-only parts. This will then allow to move the TP agent to libtorch (because libtorch is split into a CPU and a CUDA part; now it's in libtorch_python), although unfortunately other conditions need to also be met for this to happen.

The only instance where we had CPU and CUDA logic within the same code, guarded by `#ifdef USE_CUDA`, is the serialization/deserialization code. I'm thus introducing a sort-of registry in order to "decentralize it". It's not a c10::Registry, because that's overkill (it uses an unordered_map, with strings as keys): here we can just use an array with integers as "keys".

Differential Revision: [D28796428](https://our.internmc.facebook.com/intern/diff/D28796428/)

[ghstack-poisoned]

lw committed:
Update on "Make CUDA serde support for TP agent pluggable"


This is an experiment. The end goal is to separate the CUDA-specific aspects of the TensorPipe agent so that they can be plugged "on top" of the CPU-only parts. This will then allow to move the TP agent to libtorch (because libtorch is split into a CPU and a CUDA part; now it's in libtorch_python), although unfortunately other conditions need to also be met for this to happen.

The only instance where we had CPU and CUDA logic within the same code, guarded by `#ifdef USE_CUDA`, is the serialization/deserialization code. I'm thus introducing a sort-of registry in order to "decentralize it". It's not a c10::Registry, because that's overkill (it uses an unordered_map, with strings as keys): here we can just use an array with integers as "keys".

Differential Revision: [D28796428](https://our.internmc.facebook.com/intern/diff/D28796428/)

[ghstack-poisoned]

lw committed:
Update on "Make CUDA serde support for TP agent pluggable"


This is an experiment. The end goal is to separate the CUDA-specific aspects of the TensorPipe agent so that they can be plugged "on top" of the CPU-only parts. This will then allow to move the TP agent to libtorch (because libtorch is split into a CPU and a CUDA part; now it's in libtorch_python), although unfortunately other conditions need to also be met for this to happen.

The only instance where we had CPU and CUDA logic within the same code, guarded by `#ifdef USE_CUDA`, is the serialization/deserialization code. I'm thus introducing a sort-of registry in order to "decentralize it". It's not a c10::Registry, because that's overkill (it uses an unordered_map, with strings as keys): here we can just use an array with integers as "keys".

Differential Revision: [D28796428](https://our.internmc.facebook.com/intern/diff/D28796428/)

[ghstack-poisoned]

lw committed:
Update on "Make CUDA serde support for TP agent pluggable"


This is an experiment. The end goal is to separate the CUDA-specific aspects of the TensorPipe agent so that they can be plugged "on top" of the CPU-only parts. This will then allow to move the TP agent to libtorch (because libtorch is split into a CPU and a CUDA part; now it's in libtorch_python), although unfortunately other conditions need to also be met for this to happen.

The only instance where we had CPU and CUDA logic within the same code, guarded by `#ifdef USE_CUDA`, is the serialization/deserialization code. I'm thus introducing a sort-of registry in order to "decentralize it". It's not a c10::Registry, because that's overkill (it uses an unordered_map, with strings as keys): here we can just use an array with integers as "keys".

Differential Revision: [D28796428](https://our.internmc.facebook.com/intern/diff/D28796428/)

[ghstack-poisoned]

lw committed:
Update on "Make CUDA serde support for TP agent pluggable"


This is an experiment. The end goal is to separate the CUDA-specific aspects of the TensorPipe agent so that they can be plugged "on top" of the CPU-only parts. This will then allow to move the TP agent to libtorch (because libtorch is split into a CPU and a CUDA part; now it's in libtorch_python), although unfortunately other conditions need to also be met for this to happen.

The only instance where we had CPU and CUDA logic within the same code, guarded by `#ifdef USE_CUDA`, is the serialization/deserialization code. I'm thus introducing a sort-of registry in order to "decentralize it". It's not a c10::Registry, because that's overkill (it uses an unordered_map, with strings as keys): here we can just use an array with integers as "keys".

Differential Revision: [D28796428](https://our.internmc.facebook.com/intern/diff/D28796428/)

[ghstack-poisoned]

lw committed:
Update on "Make CUDA serde support for TP agent pluggable"


This is an experiment. The end goal is to separate the CUDA-specific aspects of the TensorPipe agent so that they can be plugged "on top" of the CPU-only parts. This will then allow to move the TP agent to libtorch (because libtorch is split into a CPU and a CUDA part; now it's in libtorch_python), although unfortunately other conditions need to also be met for this to happen.

The only instance where we had CPU and CUDA logic within the same code, guarded by `#ifdef USE_CUDA`, is the serialization/deserialization code. I'm thus introducing a sort-of registry in order to "decentralize it". It's not a c10::Registry, because that's overkill (it uses an unordered_map, with strings as keys): here we can just use an array with integers as "keys".

Differential Revision: [D28796428](https://our.internmc.facebook.com/intern/diff/D28796428/)

[ghstack-poisoned]

lw committed:
Update on "Make CUDA serde support for TP agent pluggable"


This is an experiment. The end goal is to separate the CUDA-specific aspects of the TensorPipe agent so that they can be plugged "on top" of the CPU-only parts. This will then allow to move the TP agent to libtorch (because libtorch is split into a CPU and a CUDA part; now it's in libtorch_python), although unfortunately other conditions need to also be met for this to happen.

The only instance where we had CPU and CUDA logic within the same code, guarded by `#ifdef USE_CUDA`, is the serialization/deserialization code. I'm thus introducing a sort-of registry in order to "decentralize it". It's not a c10::Registry, because that's overkill (it uses an unordered_map, with strings as keys): here we can just use an array with integers as "keys".

Differential Revision: [D28796428](https://our.internmc.facebook.com/intern/diff/D28796428/)

[ghstack-poisoned]

lw committed:
Update on "Make CUDA serde support for TP agent pluggable"


This is an experiment. The end goal is to separate the CUDA-specific aspects of the TensorPipe agent so that they can be plugged "on top" of the CPU-only parts. This will then allow to move the TP agent to libtorch (because libtorch is split into a CPU and a CUDA part; now it's in libtorch_python), although unfortunately other conditions need to also be met for this to happen.

The only instance where we had CPU and CUDA logic within the same code, guarded by `#ifdef USE_CUDA`, is the serialization/deserialization code. I'm thus introducing a sort-of registry in order to "decentralize it". It's not a c10::Registry, because that's overkill (it uses an unordered_map, with strings as keys): here we can just use an array with integers as "keys".

Differential Revision: [D28796428](https://our.internmc.facebook.com/intern/diff/D28796428/)

[ghstack-poisoned]

lw committed:
Update on "Make CUDA serde support for TP agent pluggable"


This is an experiment. The end goal is to separate the CUDA-specific aspects of the TensorPipe agent so that they can be plugged "on top" of the CPU-only parts. This will then allow to move the TP agent to libtorch (because libtorch is split into a CPU and a CUDA part; now it's in libtorch_python), although unfortunately other conditions need to also be met for this to happen.

The only instance where we had CPU and CUDA logic within the same code, guarded by `#ifdef USE_CUDA`, is the serialization/deserialization code. I'm thus introducing a sort-of registry in order to "decentralize it". It's not a c10::Registry, because that's overkill (it uses an unordered_map, with strings as keys): here we can just use an array with integers as "keys".

Differential Revision: [D28796428](https://our.internmc.facebook.com/intern/diff/D28796428/)

[ghstack-poisoned]

lw committed:
Update on "Make CUDA serde support for TP agent pluggable"


This is an experiment. The end goal is to separate the CUDA-specific aspects of the TensorPipe agent so that they can be plugged "on top" of the CPU-only parts. This will then allow to move the TP agent to libtorch (because libtorch is split into a CPU and a CUDA part; now it's in libtorch_python), although unfortunately other conditions need to also be met for this to happen.

The only instance where we had CPU and CUDA logic within the same code, guarded by `#ifdef USE_CUDA`, is the serialization/deserialization code. I'm thus introducing a sort-of registry in order to "decentralize it". It's not a c10::Registry, because that's overkill (it uses an unordered_map, with strings as keys): here we can just use an array with integers as "keys".

Differential Revision: [D28796428](https://our.internmc.facebook.com/intern/diff/D28796428/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_utils.cpp:
```diff
@@ -54,6 +179,18 @@ const c10::Stream& getStreamForDevice(
+    std::atomic<const TensorpipeDeviceTypeConverter*>,
+    static_cast<size_t>(DeviceType::COMPILE_TIME_MAX_DEVICE_TYPES)>
+    // NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+    device_type_converter_registry;
```
curious, since this is statically initialized and done in order, why does it still need to be atomic? What type of concurrency are we expecting?

lw replied:
It was done this way in the code I copied this from, and this was the original motivation:
https://github.com/pytorch/pytorch/blob/ff15d93b88c0e1e32d2553f272bc654d3b00606f/c10/core/impl/DeviceGuardImplInterface.h#L270-L275

I didn't want to copy all the comments/docstrings so I just mentioned where this code was copied from hoping that readers would then go check there in case of questions. Does that make sense?

mrshenli replied:
Yep, that makes sense to me.

> An unsynchronized implementation probably is OK too, but I didn't want  to prove that we never read from device_guard_impl_registry at the same time some registration is occurring.

I still wanna understand our expectations. Would I be correct if I assume in our case, it will be "A unsynchronized implementation is also OK, as TensorPipeAgent is the only owner and all initialization is done in order. The only reason we add atomic here is because this doesn't cost us anything on reads in X86."

lw replied:
I must admit I'm not sure. I suspect the main "failure" scenario is as follows: a program starts and loads `libtorch` (i.e., the CPU-only version) because that's what happens by default. Later on, one thread creates a CUDA tensor for the first time, and this triggers `libtorch_cuda` to be loaded, because it's lazy-loaded when needed. At this time, the CUDA DeviceTypeConverter is created and inserted into that array. However at this point there _could_ already be other threads that are trying to read from that array. This is _unlikely_ but technically possible I guess.

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_utils.h:
```diff
@@ -23,6 +23,54 @@ const c10::Stream& getStreamForDevice(
+class TensorpipeDeviceTypeConverter {
+ public:
+  // Ideally we'd want this to also return a tensorpipe::Message::Tensor object
+  // but we cannot forward-declare that class (because it's nested), and we
```
It's indeed a little weird to add tensors to message and then use `TORCH_INTERNAL_ASSERT` to make sure that the tensor is indeed inserted. 

Do we expect other files to access this `TensorpipeDeviceTypeConverter`? If not, maybe it can be completed implemented in the `.cpp` file and directly using `tensorpipe::Message::Tensor` there?

lw replied:
I should have made it clearer but, yes, in the next PR (#59377) I am moving the CUDA subclass of this to a separate file, thus the abstract base class must be defined in a header.

lw committed:
Update on "Make CUDA serde support for TP agent pluggable"


This is an experiment. The end goal is to separate the CUDA-specific aspects of the TensorPipe agent so that they can be plugged "on top" of the CPU-only parts. This will then allow to move the TP agent to libtorch (because libtorch is split into a CPU and a CUDA part; now it's in libtorch_python), although unfortunately other conditions need to also be met for this to happen.

The only instance where we had CPU and CUDA logic within the same code, guarded by `#ifdef USE_CUDA`, is the serialization/deserialization code. I'm thus introducing a sort-of registry in order to "decentralize it". It's not a c10::Registry, because that's overkill (it uses an unordered_map, with strings as keys): here we can just use an array with integers as "keys".

Differential Revision: [D28796428](https://our.internmc.facebook.com/intern/diff/D28796428/)

[ghstack-poisoned]
